program/rust/src/time_machine_types.rs

```diff
@@ -3,56 +3,147 @@ use crate::c_oracle_header::{
+    /// Space for more publishers
     pub extra_publisher_space: [u8; EXTRA_PUBLISHER_SPACE as usize],
-    //TimeMachine
-    pub time_machine:          TimeMachineWrapper,
+    // TimeMachine
+    pub time_machine:          SmaTracker<NUM_BUCKETS_THIRTY_MIN>,
 }
-impl PriceAccountWrapper {
-    pub fn initialize_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+
+
+#[derive(Debug)]
+pub struct DataPoint {
+    pub last_two_timestamps: (i64, i64),
+    pub slot_gap:            u64,
+    pub price:               i64,
+}
+
+impl PriceAccountExtended {
+    pub fn initialize(&mut self) -> Result<(), OracleError> {
+        self.time_machine
+            .initialize(THIRTY_MINUTES, PC_MAX_SEND_LATENCY.into());
         Ok(())
     }
 
-    pub fn add_price_to_time_machine(&mut self) -> Result<(), OracleError> {
-        // TO DO
+    pub fn add_datapoint(&mut self) -> Result<(), OracleError> {
+        self.time_machine.add_datapoint( &DataPoint{
+            last_two_timestamps : (self.price_data.prev_timestamp_, self.price_data.timestamp_),
+            slot_gap : (self.price_data.last_slot_ - self.price_data.prev_slot_),
+            price: self.price_data.prev_price_ /2  + self.price_data.agg_.price_ / 2 + (self.price_data.prev_price_ % 2) * (self.price_data.agg_.price_ % 2), // Hack to avoid overflow?
+        }
+        )?;
         Ok(())
     }
 }
 
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+pub struct SmaTracker<const NUM_BUCKETS: usize> {
+    pub threshold:                     u64,
+    pub granularity:                   i64,
+    pub current_epoch_numerator:       i128,
+    pub current_epoch_denominator:     u64,
+    pub current_epoch_is_valid:        bool,
+    pub running_sum_of_price_averages: [i128; NUM_BUCKETS],
+    pub running_valid_epoch_counter:   [u64; NUM_BUCKETS],
+}
+
+impl<const NUM_ENTRIES: usize> SmaTracker<NUM_ENTRIES> {
+    pub fn time_to_epoch(&self, time: i64) -> Result<usize, OracleError> {
+        try_convert::<i64, usize>(time / self.granularity)
+    }
+
+    pub fn initialize(&mut self, granularity: i64, threshold: u64) {
+        self.threshold = threshold;
+        self.granularity = granularity;
+        self.current_epoch_is_valid = true;
+    }
+
+    pub fn add_datapoint(&mut self, datapoint: &DataPoint) -> Result<(), OracleError> {
+        let epoch_0 = self.time_to_epoch(datapoint.last_two_timestamps.0)?;
+        let epoch_1 = self.time_to_epoch(datapoint.last_two_timestamps.1)?;
+
+        let datapoint_numerator = i128::from(datapoint.slot_gap) * i128::from(datapoint.price); //Can't overflow because u64::MAX * i64::MAX = i28::MAX
+        let datapoint_denominator = datapoint.slot_gap;
+
+        // Can't overflow, it's always smaller than the current solana slot
+        self.current_epoch_denominator += datapoint_denominator;
+        // Can't overflow, it's always smaller than u64::MAX * i64::MAX = i28::MAX
+        self.current_epoch_numerator += datapoint_numerator;
```

Comment:
not sure the overflow comment is true. `datapoint_numerator` could be `== i128::MAX` right? If I add up a few data points into `current_epoch_numerator`, then I think this can overflow?

Reply:
Added a comment with better math. Basically :
`current_epoch_numerator = slot_gap1 * price1 + slot_gap2 * price2 + slot_gap3 * price3 <= (slot_gap1 + slot_gap2 + slot_gap3) * i64::MAX <= u64::MAX * i64::MAX